lib/sproxyd.js

```diff
@@ -287,79 +287,76 @@ class SproxydClient {
-            log);
+        const req = this._createRequestHeader(method, headers, newKey, params, log);
+        const logInfo = {
+            component: 'sproxydclient',
+            method: '_handleRequest',
+            host,
+            key: newKey,
+            contentLength: size,
+        };
 
         if (stream) {
-            let isProcessingPUT = false;
+            let streamingStarted = false;
             let voluntaryAbort = false;
             headers['content-length'] = size;
 
             const request = _createRequest(req, log, (err, response) => {
                 if (err) {
-                    if (isProcessingPUT || voluntaryAbort) {
+                    if (streamingStarted || voluntaryAbort) {
                         err.retryable = false;
                     } else {
                         err.retryable = true;
                     }
                     if (!voluntaryAbort) {
-                        log.error('putting chunk to sproxyd', { host, key: newKey, error: err });
+                        logInfo.error = err;
+                        log.error('putting chunk to sproxyd', logInfo);
                     }
                     return callback(err);
                 }
                 // We return the key
-                log.debug('stored to sproxyd', {
-                    host,
-                    key: newKey,
-                    statusCode: response.statusCode,
-                });
+                logInfo.statusCode = response.statusCode;
+                log.debug('stored to sproxyd', logInfo);
                 return callback(null, newKey);
             });
-
-            request.on('finish', () => {
-                isProcessingPUT = false;
-                log.debug('finished sending PUT chunks to sproxyd', {
-                    component: 'sproxydclient',
-                    method: '_handleRequest',
-                    contentLength: size,
-                });
-            });
-
-            stream.on('error', err => {
-                isProcessingPUT = false;
-                log.error('error from readable stream', {
-                    error: err.message,
-                    method: '_handleRequest',
-                    component: 'sproxydclient',
+    
+            const startPayloadStreaming = () => {
+                streamingStarted = true;
+                stream.pipe(request);
+                finished(stream, err => {
+                    if (err) {
+                        log.trace('readable stream aborted', logInfo);
+                        request.abort();
+                        voluntaryAbort = true;
+                    } else {
+                        log.trace('readable stream finished normally', logInfo);
+                    }
                 });
+            };
+
+            // Only start streaming when connection is guaranteed: either on a
+            // new connection or a reused one.
```

Comment:
I am wondering why we do not pipe "ASAP" : waiting for the socket to be connected should be handled automatically by pipe i think...